support tenants

LizardWizzard opened the pull request:
This adds support for tenants. Support consists of:

- [x] existing cli commands 
- [x] pageserver data directory layout and existing psql commands
- [x] wal acceptor
- [x] new command for dynamic tenant creation without pageserver reload
- [x] keeping tenant name in compute postgres config so it can be used during connections to pageserver/safekeeper (done but no pr to zenith/postgres yet)

hlinnaka commented:
Nice!

What format does the tenant name have? Can we have the legendary `; DROP TABLE "COMPANIES";-- LTD` corporation (https://news.ycombinator.com/item?id=27815396) as a tenant? I'd suggest using a system-generated tenant ID everywhere, and only map that to a human-readable name in the user interface.

Seems like this should have documentation or comments somewhere to explain what it means to have multiple tenants in the page server, but I'm not sure where exactly that belongs...

LizardWizzard commented:
For now tenant name is allowed to be a sequence of any number of non space symbols, it's constrained by regexes only, for example [here](https://github.com/zenithdb/zenith/pull/318/files#diff-729d00406745f4d630d7a7a49bcc535a9c05b96f0916a65a2efa770a899c1359R371) 

Agree that same indirection as with the branch/timeline_id is better. Will transform tenant_name to tenant_id. Am I correct that page server should not know about human readable names and understand only ids in commands? So mapping to human readable format and backwards is performed in the ui.

Probably for now docs can go into pageserver's readme (it's not ideal since tenants appear not only in pageserver) or in the form of an RFC

LizardWizzard commented:
this is ready for review, corresponding pr to postgres: https://github.com/zenithdb/postgres/pull/59

lubennikovaav commented on pageserver/src/page_cache.rs:
```diff
@@ -7,29 +7,46 @@ use crate::object_repository::ObjectRepository;
-    let repo = ObjectRepository::new(conf, Arc::new(obj_store), Arc::new(walredo_mgr));
-
-    *m = Some(Arc::new(repo));
+    for dir_entry in fs::read_dir(conf.tenants_path()).unwrap() {
```
Handle error here.
If `zenith init` command was called without `--create-tenant`, `zenith start` crashes.

LizardWizzard replied:
Here I think it is better to add creation of `tenants_dir` to `init_pageserver` so here will be just empty list and no error. Is it ok?

lubennikovaav replied:
Agree. We should be able to start zenith even if no tenant exists.

lubennikovaav commented on zenith/src/main.rs:
```diff
@@ -319,23 +346,43 @@ fn get_branch_infos(env: &local_env::LocalEnv) -> Result<HashMap<ZTimelineId, Br
 
+fn handle_tenant(tenant_match: &ArgMatches, env: &local_env::LocalEnv) -> Result<()> {
+    let pageserver = PageServerNode::from_env(&env);
+    if let Some(tenantid) = tenant_match.value_of("tenantid") {
+        pageserver.tenant_create(&ZTenantId::from_str(tenantid)?)?
```
I would expect that if `tenantid` is not provided, pageserver will generate some random id.
Just like in `zenith init --create-tenant`.

lubennikovaav replied:
Also, let's print a message about operation success.

LizardWizzard replied:
I've added tenant autogeneration to cli instead of pageserver. IMO it is more suitable because calling code almost always need to keep track of newly created tenant

lubennikovaav commented on zenith/src/main.rs:
```diff
@@ -319,23 +346,43 @@ fn get_branch_infos(env: &local_env::LocalEnv) -> Result<HashMap<ZTimelineId, Br
+    if let Some(tenantid) = tenant_match.value_of("tenantid") {
+        pageserver.tenant_create(&ZTenantId::from_str(tenantid)?)?
+    } else {
+        for tenant in pageserver.tenants_list()? {
+            println!("{}", tenant);
+        }
```
Do we want to expose tenats list in CLI?
If so, let's be consistent with other commands (i.e. `pg list`) and introduce `zenith tenant list`

LizardWizzard replied:
It was the original idea, also it is useful in tests.

`zenith tenant` is actually consistent with `zenith branch`, the logic is equivalent. When there is no argument then display the list and if there is an argument then create the entity. So probably if we want to unify this we need to change `zenith branch` also. Is it suitable here? Or I can change tenant to match pg behaviour and fix branch separately? Which one is preferable?

kelvich replied:
IMO we would change `zenith branch` to smth like `zenith branch list` because we can't really add any subcommands there -- any subcommand would be treated as a branch name. So +1 on sticking with `pg list` syntax.

lubennikovaav reviewed (commented):
1. bump vendor/postgres in this branch to run CI with the recent code.

2. Let's add at least one basic python test for this feature.

3.. Please add some docs/multitenancy.md which will describe the idea and usage of this feature.
The very first example that I've tried hasn't succeeded.

```./target/debug/zenith init --create-tenant

initializing tenantid ac0350740fe19c88279f5dd399a38810
pageserver init succeeded

./target/debug/zenith start
Starting pageserver at 'localhost:64000' in .zenith
Pageserver started

./target/debug/zenith pg start  ac0350740fe19c88279f5dd399a38810 main
Starting postgres on timeline main...
Extracting base backup to create postgres instance: path=.zenith/pgdatadirs/main port=55432
Starting postgres node at 'host=127.0.0.1 port=55432 user=anastasia'
waiting for server to start.... done
server started

./target/debug/zenith tenant ac0350740fe19c88279f5dd399a38811

./target/debug/zenith pg start  ac0350740fe19c88279f5dd399a38811 main
Starting postgres on timeline main...
Starting postgres node at 'host=127.0.0.1 port=55432 user=anastasia'
pg_ctl: another server might be running; trying to start server anyway
waiting for server to start.... stopped waiting
pg_ctl: could not start server
Examine the log output.
pg operation failed: pg_ctl failed
```

LizardWizzard commented:
for python smoke tests see [here](https://github.com/zenithdb/zenith/pull/318/files#diff-85440eac3bb111914ec6d1e6029a82c37f6dc7f303dda8a33d9c7d56e4ffc8b9) and [here](https://github.com/zenithdb/zenith/pull/318/files#diff-407c5aaf3328f9a7b9c06ca62b8781c48fa2ee45fbd3d14e2cc4569ec05c136d) also I think that it is a good idea to add your example as a separate test case

lubennikovaav commented on pageserver/src/page_service.rs:
```diff
@@ -328,84 +333,102 @@ impl postgres_backend::Handler for PageServerHandler {
+                "invalid param number for pagestream command"
+            );
+
+            let timelineid = ZTimelineId::from_str(params[0])?;
+            let tenantid = ZTenantId::from_str(params[1])?;
```
Let's change the order of parameters to be consistent with other commands.

Now we have: 
```
pageserver timelineid tenantid
callmemaybe tenantid timelineid
```

lubennikovaav commented on pageserver/src/page_service.rs:
```diff
@@ -485,54 +509,99 @@ impl postgres_backend::Handler for PageServerHandler {
-            .starts_with(b"identify_system")
+            .starts_with("identify_system")
         {
+            // NOTE identify system is a postgres own thing so it is better to preserve compatibility however,
+            // for executing this the knowledge of tenant id is required, so it cannot be passed directly in query itself
+            // as discussed with @kelvich this is currently not in use and planned for removal in future during introduction
+            // of new replication flow which will replace callmemaybe machinery
+            // so here for simplicity we add tenant name to query.
+            // When needed it can be passed in connection string and extracted from startup's packet FeStartupMessage.params
```
@kelvich, can we remove this code? I don't see if it is used anywhere.

kelvich replied:
Yeah, I think we can just delete that code

lubennikovaav replied:
Done in #356

lubennikovaav reviewed (commented):
Please run `cargo fmt` on your branch

LizardWizzard commented on pageserver/src/page_service.rs:
```diff
@@ -328,84 +333,102 @@ impl postgres_backend::Handler for PageServerHandler {
+            let tenantid = ZTenantId::from_str(params[0])?;
+            let timelineid = ZTimelineId::from_str(params[1])?;
+
+            // TODO are there any tests with lsn option?
+            let lsn = if params.len() == 3 {
+                Some(Lsn::from_str(params[2])?)
```
@kelvich are you aware of any usages (maybe future ones) of basebackup with lsn param? Can it be removed now and added back if needed?

kelvich replied:
Yeah, there is no current basebackup@lsn users that I am aware of. Though functionality itself sounds useful for read-only replicas pinned to some LSN. So there are some future use cases, but if this code is standing in your way -- feel free to delete it. We can add that later with tests.

LizardWizzard commented:
Updates:
* Found that create tenant command in page service didnt update global repository map, fixed
* Added level of indirection for tenants in `pgdatadirs` so your example works correctly now
* Added one more test for tenants, see `test_tenants.py`
* Changed handling of command argumets of tenant command to match `zenith pg` behaviour
* Tenant create without tenant id now generates one by default (on the cli side)
* Added docs page under `docs/multitenancy.md`

kelvich commented on docs/multitenancy.md:
````diff
@@ -0,0 +1,60 @@
+
+### Overview
+
+Zenith supports multitenancy. One pagegserver can serve multiple tenants at once. Tenants can be managed via zenith CLI. During page server setup tenant can be created using ```zenith init --create-tenant``` Also tenants can be added into the system on the fly without pageserver restart. This can be done using the following cli command: ```zenith tenant create``` Tenants use random identifiers wh ...
````
typo

kelvich commented on docs/multitenancy.md:
````diff
@@ -0,0 +1,60 @@
+
+// lists branches for <tenantid>
+zenith branch ee6016ec31116c1b7c33dfdfca38892f
+```
````
I think it is better to have default tenant and don't require tenant_id. If somebody wants to access non-default tenant he may specify --tenant

lubennikovaav commented:
> For now tenant name is allowed to be a sequence of any number of non space symbols, it's constrained by regexes only, for example [here](https://github.com/zenithdb/zenith/pull/318/files#diff-729d00406745f4d630d7a7a49bcc535a9c05b96f0916a65a2efa770a899c1359R371)
> 
> Agree that same indirection as with the branch/timeline_id is better. Will transform tenant_name to tenant_id. Am I correct that page server should not know about human readable names and understand only ids in commands? So mapping to human readable format and backwards is performed in the UI.
> 

I suggest that we do that now. I've tested the PR a bit and found the need to copy tenantid to every command quite annoying.
CLI seems to be the right place to maintain the mapping, so that pageserver only operates with ids and the user only uses human-readable names. Maybe we can expose the mapping in `zenith tenant list`, but I am not sure if it is needed for anything but debugging.

And also we can introduce a default name for the initial tenant (the one created via `zenith init --create-tenant`.
It will allow us to simplify tests.

kelvich commented:
I would rather not bother with tenant name at all. Let always create default tenant during `zenith init` store it in config file and use command without necessity to specify tenant_id.

kelvich commented:
Generally I like the state of this PR, let's rework CLI part and merge it

lubennikovaav commented on docs/multitenancy.md:
````diff
@@ -0,0 +1,60 @@
+
+### Tenants in other commands
+
+Frequently tenant id apper in other commands, for example in branch create command: ```zenith branch ee6016ec31116c1b7c33dfdfca38891f new_one main``` creates branch `new_one` forked from `main` for tenant `ee6016ec31116c1b7c33dfdfca38891f`. This is also the case for internal page server interface used for communication with compute nodes. So here for example `pagestream` command will also require ...
````
apper -> appear

LizardWizzard commented:
Updates:
* CLI now creates tenant in pageserver without opting out, however cli of pageserver itself support init without tenant creation
* after `zenith init` CLI saves initial tenant in config, and then uses it as a fallback value in other commands. Other commands now support optional argument --tenantid=<tenantid>
* test fixtures enchanced to remember initial tenant (and here CLI fallback also works) so it is not required to carry initial tenant id to each call of e.g. Postgres.create
* Rebased on main. (This includes removal of identify_system paegserver command)

kelvich commented:
Overall look good. One weird scenario is `pg list`:
```
> ./target/debug/zenith pg list                                            
BRANCH	ADDRESS		LSN		STATUS
main	127.0.0.1:55432	0/16B91C0	running
main	127.0.0.1:55433	?	running
```

May be just add --tenantid as with create/start/stop/branch?
